libbpf-tools/cpufreq.c

```diff
@@ -0,0 +1,254 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Wenbo Zhang
+//
+// Based on cpufreq(8) from BPF-Perf-Tools-Book by Brendan Gregg.
+// 10-OCT-2020   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <linux/perf_event.h>
+#include <asm/unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "cpufreq.h"
+#include "cpufreq.skel.h"
+#include "trace_helpers.h"
+
+static struct env {
+	int duration;
+	int freq;
+	bool verbose;
+} env = {
+	.duration = -1,
+	.freq = 99,
+};
+
+const char *argp_program_version = "cpufreq 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+const char argp_program_doc[] =
+"Sampling CPU freq system-wide & by process. Ctrl-C to end.\n"
+"\n"
+"USAGE: cpufreq [--help] [-d DURATION] [freq]\n"
```

Comment:
it's weird do have duration as -d argument, but freq as positional, tbh. I'd just make freq as -f argument or something along those lines.

Reply:
Fixed, thanks!